Add friend tree capabilities to spark

vepadulano opened the pull request:
This is a barebones example of implementation to deal with
`TChain`s that have friend trees. The implementation follows the reasoning of
`GetFriendInfo` and `MakeChain` functions of the [TTreeProcessorMT](https://root.cern.ch/doc/master/classROOT_1_1TTreeProcessorMT.html) ROOT class.


The `RDataFrame` class is already capable of retrieveng the filenames of
the TTree/Tchain object passed as an argument to `PyRDF.RDataFrame()`.

Now it is also capable of retrieving treenames and filenames of friend trees,
if there are any. For the moment, the friend trees list is retrieved via
`TTree.GetListOfFriends()` and for each element in the list a tuple with
(treename, filename) is appended to a list.

During the map phase, in `Dist.py`, the list of tuples is retrieved and
added as attribute to each `Range` instance of the distributed execution,
as was previously done for the list of filenames of the `TTree`s in the
chain.

Finally, after reconstructing the `TChain`, the friend are added to the chain
via `TChain.AddFriend()`.

A test has been modified to also include the new attribute of the `Range` class

vepadulano committed:
[WIP] Add friend tree capabilities to spark

This is a barebones example of implementation to deal with
`TChain`s that have friend trees. The implementation follows the reasoning of
`GetFriendInfo` and `MakeChain` functions of the `TTreeProcessorMT` ROOT class.
https://root.cern.ch/doc/master/classROOT_1_1TTreeProcessorMT.html

The `RDataFrame` class is already capable of retrieveng the filenames of
the TTree/Tchain object passed as an argument to `PyRDF.RDataFrame()`.

Now it is also capable of retrieving treenames and filenames of friend trees,
if there are any. For the moment, the friend trees list is retrieved via
`TTree.GetListOfFriends()` and for each element in the list a tuple with
(treename, filename) is appended to a list.

During the map phase, in `Dist.py`, the list of tuples is retrieved and
added as attribute to each `Range` instance of the distributed execution,
as was previously done for the list of filenames of the `TTree`s in the
chain.

Finally, after reconstructing the `TChain`, the friend are added to the chain
via `TChain.AddFriend()`.

A test has been modified to also include the new attribute of the `Range` class

JavierCVilla commented on PyRDF/RDataFrame.py:
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
         # RDataFrame may have been created with no input files
         return None
 
+    def get_input_friend_treenames_and_filenames(self):
```
Even though I can understand why you put it here, I would not implement this method in `RDataFrame.py` but in `Dist.py`. 

While the rest of `get` methods (`get_inputfiles`, `get_treename`, `get_branches` ... ) need to check the parameters passed to the `RDataFrame` constructor in order to access to the required information (input files, treename, branches,...), this method could be directly applied to a `TTree/TChain` as it is done in the ROOT implementation. Indeed, this method will be only called from a `Dist` class or subclass.

JavierCVilla replied:
Also I would use a shorter name, maybe the same as in the ROOT implementation: `get_friend_info`.

JavierCVilla commented on PyRDF/RDataFrame.py:
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
+
+        """
+        first_arg = self.args[0]
+        if isinstance(first_arg, ROOT.TChain) and first_arg.GetListOfFriends():
```
Moving the method to `Dist` and getting a `TTree` as an argument, we can even remove these checks.

JavierCVilla commented on PyRDF/RDataFrame.py:
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
+            # Extract file names from a given TChain
+            chain = first_arg
+            return [
+                (friendElem.GetName(), friendElem.GetTitle())
```
For the result we may want to be more explicit and use either namedtuples or a custom object to represent the `FriendInfo`.

JavierCVilla commented on PyRDF/backend/Dist.py:
```diff
@@ -209,7 +211,8 @@ def _get_clustered_ranges(self, nentries, npartitions, treename, filelist):
                                 end - offset_first_cluster,
-                                range_files))
+                                range_files,
+                                friend_filelist))
```
`range_friend_files`

vepadulano replied:
I am in doubt about this one particularly. Thing is that `range_files` is created after retrieving the file name for that particular range of entries, but that doesn't automatically mean that for the same range of entries we would have the same friend file names, right? 
For the moment I am just passing all the filenames (and treenames actually) of the friend trees

JavierCVilla replied:
Right, this might require further investigation (also to see how it is currently implemented in root), as clusters in both sets of files may not be the same. 

`range_files` is added to the list of ranges for debugging purposes and to create `TChain`'s with a reduced size so we do not add unused files. At some point we would need to do the same for friend trees/chains but for now we can focus on the creation of the friend chain and add this optimization later on.

JavierCVilla commented on PyRDF/backend/Dist.py:
```diff
@@ -306,6 +315,10 @@ def mapper(current_range):
 
+                if current_range.friend_filelist:
+                    for f in current_range.friend_filelist:
+                        chain.AddFriend(f[0], f[1])
```
This assumes that the friend trees only have one file, but it may also be a chain composed by many files.

First we need to recreate the same chain before adding it as friend, as in the root implementation.

Additionally, we cannot just add the friend, since we need to set the number of entries to be read, as we do for `chain` with `chain.SetCacheEntryRange`.

JavierCVilla reviewed (changes requested):
Thank you @vepadulano for the quick PR after spotting the issue!

Proposed changes go on the right direction, however we need to rethink a bit the support for friend trees and reuse as much code as possible from the ROOT implementation as it solves the same issue.

JavierCVilla commented on tests/unit/backend/test_dist.py:
```diff
@@ -285,6 +285,7 @@ def test_buildranges_with_clustered_ranges(self):
         # Mock attributes accessed through self. inside build_ranges
         backend.treename = "myTree"
         backend.files = "tests/unit/backend/1000clusters.root"
+        backend.friend_filelist = None
```
Before I forget, we need a test with friend trees.

vepadulano committed:
Move friend_info to Dist and better manage friend names and filenames

Now `_get_friend_info` is a method of `Dist` class and returns a namedtuple
`FriendInfo` containing two lists. The first list holds the names of the friend trees,
while the second holds the corresponding filenames where the trees are stored. Each
treename can correspond to multiple file names.

During map phase, a new TChain is initialized for each treename of the friends, and all
the corresponding filenames are added to it. Then the new friend chain is cached with
the same range as the main TChain (it is supposed that the entries are in the same order).
Finally the friend chain is added to the main chain and the RDataFrame is initialized with the latter.

A new test has been added to check the functionalities of RDataFrame on Spark with friend trees. In
the test, an RDataFrame is initialized with one TChain holding a friend TChain, each having one column `x`.
One histogram is created for each of the two columns, and common statistics are retrieved from them.

vepadulano committed:
create FriendInfo class to hold friend trees information

A new class `FriendInfo` has been added to `Dist.py` to hold information
about friend trees. This class has two attributes, `friend_names` and
`friend_file_names` which are lists holding respectively the treename
and the corresponding file names for each friend tree.

In `execute` an initial check that the RDataFrame was initialized with a
`ROOT.TTree`[`TChain`] object is performed, then the information about the
friend trees is retrieved. This is made possible by the addition of a
new method `get_tree` in the `RDataFrame` class.

A couple of tests have been added to evaluate the behaviour of the new
`FriendInfo` class.

JavierCVilla commented on tests/integration/spark/test_include_headers.py:
```diff
@@ -8,11 +8,20 @@ class IncludesSparkTest(unittest.TestCase):
     Check that the required header files are properly included in Spark
     environment.
     """
+    def tear_down(self):
```
Can this be implemented with `tearDown` instead?

JavierCVilla commented on tests/integration/spark/test_include_headers.py:
```diff
@@ -8,11 +8,20 @@ class IncludesSparkTest(unittest.TestCase):
         were included using header files.
         """
+        # Stop current SparkContext if present
+        self.tear_down()
```
This would not be needed if `tead_down` is implemented as `teadDown`

JavierCVilla commented on tests/integration/spark/test_inv_mass.py:
```diff
@@ -5,9 +5,7 @@
-
-    @classmethod
-    def tearDown(cls):
+    def tear_down(self):
```
What's the reason of this change?
